### ISSUE TITLE
needless_match bad suggestion with unsize coercion

### ISSUE THREAD

**Issue by ijackson** [rust-lang/rust-clippy#8551]
### Summary

`needless_match` can suggest replacing a match statement with an arm implying a coercion, with a simple expression which no longer does the coercion.

### Lint Name

needless_match

### Reproducer

I tried this code:

```rust
trait Trait { }
struct Struct;
impl Trait for Struct { }

fn optmap(s: Option<&Struct>) -> Option<&dyn Trait> {
    match s {
        Some(s) => Some(s),
        None => None,
    }
}

fn main() {
    optmap(None);
}
```

I saw this happen:

```
nailing-cargo: invoking: cargo clippy --manifest-path=/home/ian/Rustup/Arti/experiments/Cargo.toml --locked --target-dir=target --offline
    Checking foo v0.1.0 (/home/ian/Rustup/Arti/experiments)
warning: this match expression is unnecessary
 --> src/main.rs:6:5
  |
6 | /     match s {
7 | |         Some(s) => Some(s),
8 | |         None => None,
9 | |     }
  | |_____^ help: replace it with: `s`
  |
  = note: `#[warn(clippy::needless_match)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_match

warning: `foo` (bin "foo") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
nailing-cargo: unnailed.  status 0.
```

I expected to see this happen:

Either no complaint, or, better, suggesting something like
```
fn optmap(s: Option<&Struct>) -> Option<&dyn Trait> {
    s.map(|s| s as _)
}
```

### Version

```text
rustc 1.61.0-nightly (52b34550a 2022-03-15)
binary: rustc
commit-hash: 52b34550aca5f7dd7e152f773e3ab786acb86f6f
commit-date: 2022-03-15
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
